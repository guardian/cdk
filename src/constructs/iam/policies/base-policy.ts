import { Effect, Policy, PolicyStatement } from "aws-cdk-lib/aws-iam";
import type { PolicyProps } from "aws-cdk-lib/aws-iam";
import type { GuStack } from "../../core";

export type GuPolicyProps = PolicyProps;

export type GuNoStatementsPolicyProps = Omit<GuPolicyProps, "statements">;

export class GuPolicy extends Policy {
  constructor(scope: GuStack, id: string, props: GuPolicyProps) {
    super(scope, id, props);
  }
}

export interface GuAllowPolicyProps extends GuNoStatementsPolicyProps {
  actions: string[];
  resources: string[];
}
export type GuDenyPolicyProps = GuAllowPolicyProps;

export class GuAllowPolicy extends GuPolicy {
  constructor(scope: GuStack, id: string, props: GuAllowPolicyProps) {
    super(scope, id, props);

    this.addStatements(
      new PolicyStatement({
        effect: Effect.ALLOW,
        resources: props.resources,
        actions: props.actions,
      })
    );
  }
}

export class GuDenyPolicy extends GuPolicy {
  constructor(scope: GuStack, id: string, props: GuDenyPolicyProps) {
    super(scope, id, props);

    this.addStatements(
      new PolicyStatement({
        effect: Effect.DENY,
        resources: props.resources,
        actions: props.actions,
      })
    );
  }
}
