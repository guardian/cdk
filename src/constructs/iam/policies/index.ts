export * from "./assume-role";
export * from "./anghammarad";
export * from "./base-policy";
export * from "./cloudwatch";
export * from "./describe-ec2";
export * from "./dynamodb";
export * from "./log-shipping";
export * from "./parameter-store-read";
export * from "./s3-get-object";
export * from "./s3-put-object";
export * from "./ses";
export * from "./ssm";
