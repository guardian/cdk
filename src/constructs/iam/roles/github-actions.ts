import { FederatedPrincipal } from "@aws-cdk/aws-iam";
import { CfnOutput, CfnResource } from "@aws-cdk/core";
import type { GuStack } from "../../core";
import type { GuPolicy } from "../policies";
import { GuRole } from "./roles";

const GITHUB_ACTIONS_ID_TOKEN_REQUEST_DOMAIN = "token.actions.githubusercontent.com";

/*
Thumbprint of `GITHUB_ACTIONS_ID_TOKEN_REQUEST_DOMAIN`.
See: https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc_verify-thumbprint.html
*/
const GITHUB_ACTIONS_ID_TOKEN_REQUEST_DOMAIN_THUMBPRINT = "6938fd4d98bab03faadb97b34396831e3780aea1";

export interface GuGithubActionsRepositoryCondition {
  /**
   * The GitHub organisation/user to constrain the IAM Role to.
   */
  githubOrganisation: string;

  /**
   * Repositories to constrain the IAM Role to.
   */
  repositories?: string;
}

const GuGithubActionsRepositoryCondition = {
  toString: ({ githubOrganisation, repositories }: GuGithubActionsRepositoryCondition): string =>
    `repo:${githubOrganisation}/${repositories ?? "*"}`,
};

const ALL_GUARDIAN_REPOSITORIES: GuGithubActionsRepositoryCondition = {
  githubOrganisation: "guardian",
  repositories: "*",
};

export interface GuGithubActionsRoleProps {
  /**
   * A list of IAM Policies for the GitHub Action.
   */
  policies: GuPolicy[];

  /**
   * Repositories where GitHub Actions can assumes this role.
   * Defaults to [[`ALL_GUARDIAN_REPOSITORIES`]].
   */
  condition?: GuGithubActionsRepositoryCondition;
}

/*
AWS CDK implements an OIDCProvider as a custom resource.
This requires a lambda to be deployed into the account.
As far as I can tell, the lambda is automating the setting of `ThumbprintList`, which is quite generic.
It's simpler for us to be specific here.

See:
  - https://github.com/aws/aws-cdk/blob/851c8ca9989856fa61496ff113f9cb8c66d02f3b/packages/%40aws-cdk/aws-iam/lib/oidc-provider.ts
  - https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc.html
 */
class GitHubOidcProvider extends CfnResource {
  constructor(scope: GuStack) {
    super(scope, "GithubActionsOidc", {
      type: "AWS::IAM::OIDCProvider",
      properties: {
        Url: `https://${GITHUB_ACTIONS_ID_TOKEN_REQUEST_DOMAIN}`,
        ClientIdList: ["sigstore"],
        ThumbprintList: [GITHUB_ACTIONS_ID_TOKEN_REQUEST_DOMAIN_THUMBPRINT],
      },
    });
  }
}

/**
 * A construct to create an IAM role for GitHub Actions to assume via `AssumeRoleWithWebIdentity`.
 *
 * An Output will be added to the stack with the Role's ARN for use with https://github.com/aws-actions/configure-aws-credentials.
 *
 * See:
 *   - https://github.com/aws-actions/configure-aws-credentials
 *   - https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc.html
 */
export class GuGithubActionsRole extends GuRole {
  constructor(scope: GuStack, props: GuGithubActionsRoleProps) {
    super(scope, "GithubActionsRole", {
      assumedBy: new FederatedPrincipal(
        new GitHubOidcProvider(scope).ref,
        {
          StringLike: {
            [`${GITHUB_ACTIONS_ID_TOKEN_REQUEST_DOMAIN}:sub`]: GuGithubActionsRepositoryCondition.toString(
              props.condition ?? ALL_GUARDIAN_REPOSITORIES
            ),
          },
        },
        "sts:AssumeRoleWithWebIdentity"
      ),
    });

    props.policies.forEach((p) => p.attachToRole(this));

    new CfnOutput(this, `${this.node.id}-Arn`, { value: this.roleArn });
  }
}
