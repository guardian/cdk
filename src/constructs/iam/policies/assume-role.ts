import { Effect, PolicyStatement } from "aws-cdk-lib/aws-iam";
import type { GuStack } from "../../core";
import { GuPolicy } from "./base-policy";
import type { GuNoStatementsPolicyProps } from "./base-policy";

export interface GuAssumeRolePolicyProps extends GuNoStatementsPolicyProps {
  resources: string[];
}

export const guAssumeRolePolicyStatement = (resources: string[]) =>
  new PolicyStatement({
    effect: Effect.ALLOW,
    actions: ["sts:AssumeRole"],
    resources: resources,
  });

export class GuAssumeRolePolicy extends GuPolicy {
  constructor(scope: GuStack, id: string, props: GuAssumeRolePolicyProps) {
    super(scope, id, {
      statements: [guAssumeRolePolicyStatement(props.resources)],
      ...props,
    });
  }
}
