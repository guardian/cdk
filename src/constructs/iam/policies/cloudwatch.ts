import { GuAllowPolicy } from "./base-policy";
import type { GuStack } from "../../core";
import type { GuAllowPolicyProps } from "./base-policy";

abstract class GuCloudwatchPolicy extends GuAllowPolicy {
  protected constructor(scope: GuStack, id: string, props: Omit<GuAllowPolicyProps, "resources">) {
    super(scope, id, {
      ...props,
      actions: props.actions.map((action) => `cloudwatch:${action}`),
      resources: ["*"],
    });
  }
}

export class GuGetCloudwatchMetricsPolicy extends GuCloudwatchPolicy {
  constructor(scope: GuStack) {
    super(scope, "GuGetCloudwatchMetricsPolicy", {
      actions: ["ListMetrics", "GetMetricData", "GetMetricStatistics", "DescribeAlarmsForMetric"],
    });
  }
}

export class GuPutCloudwatchMetricsPolicy extends GuCloudwatchPolicy {
  constructor(scope: GuStack) {
    super(scope, "GuPutCloudwatchMetricsPolicy", { actions: ["PutMetricData"] });
  }
}
