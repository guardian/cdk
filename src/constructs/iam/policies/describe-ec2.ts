import type { PolicyProps } from "@aws-cdk/aws-iam";
import { Effect, PolicyStatement } from "@aws-cdk/aws-iam";
import type { GuStack } from "../../core";
import type { GuPolicyProps } from "./base-policy";
import { GuPolicy } from "./base-policy";

export class GuDescribeEC2Policy extends GuPolicy {
  private static getDefaultProps(): PolicyProps {
    return {
      policyName: "describe-ec2-policy",
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: [
            "autoscaling:DescribeAutoScalingInstances",
            "autoscaling:DescribeAutoScalingGroups",
            "ec2:DescribeTags",
            "ec2:DescribeInstances",
          ],
          resources: ["*"],
        }),
      ],
    };
  }

  constructor(scope: GuStack, id: string = "DescribeEC2Policy", props?: GuPolicyProps) {
    super(scope, id, { ...GuDescribeEC2Policy.getDefaultProps(), ...props });
  }
}
