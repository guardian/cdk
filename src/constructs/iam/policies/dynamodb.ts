import { GuAllowPolicy } from "./base-policy";
import type { GuStack } from "../../core";
import type { GuNoStatementsPolicyProps } from "./base-policy";

interface GuDynamoDBPolicyProps {
  tableName: string;
}

interface GuDynamoDBPolicyPropsWithActions extends GuNoStatementsPolicyProps, GuDynamoDBPolicyProps {
  actions: string[];
}

abstract class GuDynamoDBPolicy extends GuAllowPolicy {
  protected constructor(scope: GuStack, id: string, props: GuDynamoDBPolicyPropsWithActions) {
    super(scope, id, {
      ...props,
      actions: props.actions.map((action) => `dynamodb:${action}`),
      resources: [`arn:aws:dynamodb:${scope.region}:${scope.account}:table/${props.tableName}`],
    });
  }
}

export class GuDynamoDBReadPolicy extends GuDynamoDBPolicy {
  constructor(scope: GuStack, id: string, props: GuDynamoDBPolicyProps) {
    super(scope, id, { ...props, actions: ["BatchGetItem", "GetItem", "Scan", "Query", "GetRecords"] });
  }
}

export class GuDynamoDBWritePolicy extends GuDynamoDBPolicy {
  constructor(scope: GuStack, id: string, props: GuDynamoDBPolicyProps) {
    super(scope, id, { ...props, actions: ["BatchWriteItem", "PutItem", "DeleteItem", "UpdateItem"] });
  }
}
