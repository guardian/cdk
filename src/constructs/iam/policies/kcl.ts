import { Effect, PolicyStatement } from "aws-cdk-lib/aws-iam";
import type { IStream } from "aws-cdk-lib/aws-kinesis";
import type { GuStack } from "../../core";
import { GuPolicy } from "./base-policy";

export interface KCLApplication {
  stream: IStream;
  applicationName: string;
}

const kinesisActions: string[] = [
  "DescribeStream",
  "DescribeStreamSummary",
  "RegisterStreamConsumer",
  "GetRecords",
  "GetShardIterator",
  "ListShards",
];

const kinesisEnhancedFanOutActions: string[] = ["SubscribeToShard", "DescribeStreamConsumer"];

const actionsOnAllTables: string[] = [
  "Scan",
  "CreateTable", // Allow KCL creating the lease, metrics, and coordinator tables - see https://github.com/guardian/cdk/pull/2578#discussion_r1954061397
  "DescribeTable",
  "GetItem",
  "PutItem",
];

const additionalLeaseTableActions: string[] = ["UpdateTable", "UpdateItem", "DeleteItem"];

/**
 * Creates an `AWS::IAM::Policy` to grant all the required permissions for the Kinesis Client Library
 * (https://github.com/awslabs/amazon-kinesis-client) - specifically for KCL v3, which requires many
 * additional permissions over KCL v2.
 *
 * @see https://docs.aws.amazon.com/streams/latest/dev/kcl-iam-permissions.html
 */
export class GuKCLPolicy extends GuPolicy {
  constructor(scope: GuStack, props: KCLApplication) {
    function allow(actionType: string, actions: string[], resources: string[]): PolicyStatement {
      return new PolicyStatement({
        effect: Effect.ALLOW,
        actions: actions.map((a) => `${actionType}:${a}`),
        resources: resources.map((r) => `arn:aws:${actionType}:${scope.region}:${scope.account}:${r}`),
      });
    }

    function allowDynamoDB(actions: string[], tableOrIndexNames: string[]): PolicyStatement {
      return allow(
        "dynamodb",
        actions,
        tableOrIndexNames.map((name) => `table/${name}`),
      );
    }

    const leaseTable = props.applicationName;
    const metricsTable = `${props.applicationName}-WorkerMetricStats`;
    const coordinatorTable = `${props.applicationName}-CoordinatorState`;

    super(scope, `GuKCLPolicy${props.applicationName}`, {
      statements: [
        allow("kinesis", kinesisActions, [`stream/${props.stream.streamName}`]),
        allow("kinesis", kinesisEnhancedFanOutActions, [`stream/${props.stream.streamName}/consumer/*`]),
        allowDynamoDB(actionsOnAllTables, [leaseTable, metricsTable, coordinatorTable]),
        allowDynamoDB(additionalLeaseTableActions, [leaseTable]),
        allowDynamoDB(["Query"], [`${leaseTable}/index/*`]),
        allow("cloudwatch", ["PutMetricData"], ["*"]),
      ],
    });
  }
}
