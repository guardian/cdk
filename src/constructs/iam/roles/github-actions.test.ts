import { Template } from "aws-cdk-lib/assertions";
import { simpleGuStackForTesting } from "../../../utils/test";
import { GuGetS3ObjectsPolicy } from "../policies";
import { GuGithubActionsRole } from "./github-actions";

describe("The GitHubActionsRole construct", () => {
  it("should create the correct resources with minimal config", () => {
    const stack = simpleGuStackForTesting();
    new GuGithubActionsRole(stack, {
      policies: [
        new GuGetS3ObjectsPolicy(stack, "GetObjects", {
          bucketName: "super-secret-stuff",
        }),
      ],
    });

    expect(Template.fromStack(stack).toJSON()).toMatchSnapshot();
  });

  it("should be possible to limit which repositories can assume the role", () => {
    const stack = simpleGuStackForTesting();
    new GuGithubActionsRole(stack, {
      policies: [
        new GuGetS3ObjectsPolicy(stack, "GetObjects", {
          bucketName: "super-secret-stuff",
        }),
      ],
      condition: {
        githubOrganisation: "guardian",
        repositories: "platform-*",
      },
    });

    Template.fromStack(stack).hasResourceProperties("AWS::IAM::Role", {
      AssumeRolePolicyDocument: {
        Statement: [
          {
            Action: "sts:AssumeRoleWithWebIdentity",
            Condition: {
              StringLike: {
                "token.actions.githubusercontent.com:sub": "repo:guardian/platform-*",
              },
            },
          },
        ],
      },
    });
  });
});
