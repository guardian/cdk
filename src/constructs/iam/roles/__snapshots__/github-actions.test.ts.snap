// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The GitHubActionsRole construct should create the correct resources with minimal config 1`] = `
Object {
  "Outputs": Object {
    "GithubActionsRoleGithubActionsRoleArnC13D9654": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "GithubActionsRoleF5CC769F",
          "Arn",
        ],
      },
    },
  },
  "Parameters": Object {
    "Stage": Object {
      "AllowedValues": Array [
        "CODE",
        "PROD",
      ],
      "Default": "CODE",
      "Description": "Stage name",
      "Type": "String",
    },
  },
  "Resources": Object {
    "GetObjectsDE66E0D1": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Resource": "arn:aws:s3:::super-secret-stuff/*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GetObjectsDE66E0D1",
        "Roles": Array [
          Object {
            "Ref": "GithubActionsRoleF5CC769F",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "GithubActionsOidc": Object {
      "Properties": Object {
        "ClientIdList": Array [
          "sigstore",
        ],
        "ThumbprintList": Array [
          "6938fd4d98bab03faadb97b34396831e3780aea1",
        ],
        "Url": "https://token.actions.githubusercontent.com",
      },
      "Type": "AWS::IAM::OIDCProvider",
    },
    "GithubActionsRoleF5CC769F": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": Object {
                "StringLike": Object {
                  "token.actions.githubusercontent.com:sub": "repo:guardian/*",
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Federated": Object {
                  "Ref": "GithubActionsOidc",
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
